internal/appsec/remoteconfig.go

```diff
@@ -41,9 +43,120 @@ func statusesFromUpdate(u remoteconfig.ProductUpdate, ack bool, err error) map[s
+		statuses = mergeMaps(statuses, a.handleASMFeatures(u))
+		delete(updates, rc.ProductASMFeatures)
+	}
+
+	var err error
+updateLoop:
+	// Process rules related updates
+	for p, u := range updates {
+		switch p {
+		case rc.ProductASMData:
+			// Merge all rules data entries together and store them as a rulesManager edit entry
+			rulesData, status := mergeRulesData(u)
+			statuses = mergeMaps(statuses, status)
+			r.addEdit("asmdata", rulesFragment{RulesData: rulesData})
+		case rc.ProductASMDD:
+			// Switch the base rules of the rulesManager if the config received through ASM_DD is valid
+			// If the config was removed, switch back to the static recommended rules
+			if len(u) > 1 { // Don't process configs if more than one is received for ASM_DD
+				log.Debug("appsec: Remote config: more than one config received for ASM_DD. Updates won't be applied")
+				err = errors.New("More than one config received for ASM_DD")
+				statuses = mergeMaps(statuses, statusesFromUpdate(u, true, err))
+				break updateLoop
+			}
+			for path, data := range u {
+				if data == nil {
+					log.Debug("appsec: Remote config: ASM_DD config removed. Switching back to default rules")
+					r.changeBase(defaultRulesFragment(), "")
+					break
+				}
+				var newBase rulesFragment
+				if err = json.Unmarshal(data, &newBase); err != nil {
+					log.Debug("appsec: Remote config: could not unmarshall ASM_DD rules: %v", err)
+					statuses[path] = genApplyStatus(true, err)
+					break updateLoop
+				}
+				log.Debug("appsec: Remote config: switching to %s as the base rules file", path)
+				r.changeBase(newBase, path)
+			}
+		case rc.ProductASM:
+			// Store each config received through ASM as an edit entry in the rulesManager
+			// Those entries will get merged together when the final rules are compiled
+			// If a config gets removed, the rulesManager edit entry gets removed as well
+			for path, data := range u {
+				log.Debug("appsec: Remote config: processing the %s ASM config", path)
+				if data == nil {
+					log.Debug("appsec: Remote config: ASM config %s was removed", path)
+					r.removeEdit(path)
+					continue
+				}
+				var f rulesFragment
+				if err = json.Unmarshal(data, &f); err != nil || !f.validate() {
+					if err == nil { //
+						err = errors.New("invalid configuration payload")
+					}
+					log.Debug("appsec: Remote config: error processing ASM config %s: %v", path, err)
+					statuses[path] = genApplyStatus(true, err)
+					break updateLoop
+				} else {
+					r.addEdit(path, f)
+				}
+			}
+		default:
+			log.Debug("appsec: Remote config: ignoring unsubscribed product %s", p)
+		}
+	}
+
+	if err != nil {
+		log.Debug("appsec: Remote config: not applying any updates because of error: %v", err)
+		return statuses
+	}
+
+	// Set all statuses to ack
+	for _, u := range updates {
+		statuses = mergeMaps(statuses, statusesFromUpdate(u, true, nil))
+	}
+
+	// Compile the final rules once all updates have been processed
```

Comment:
I think it would be quite hard to do so. Here are the cases I can think about:
- "Normal" updates -> rules have changed
- No update -> no call to this callback
- Updates but errors -> rules don't change but we've early exited
- Updates that contradict each-other (ex: 2 override entries with enable/disable of the same rule): even a deepequal can't check that since those entries are concatenated under "rules_overrides" and the WAF takes care of merging this, and trying to handle this ourselves seems quite challenging if we have to do so for every config.

What do you think?

Reply:
Yes, all good. IIRC you didn't have the early exit on errors implemented yesterday, but now you do and this is good then. Thanks